bigbang/ingress/abstract.py

```diff
@@ -0,0 +1,773 @@
+            else:
+                header = self.empty_header
+            if fields in ["body", "total"]:
+                body = self._get_body_from_html(list_name, url, soup)
+            else:
+                body = None
+        return self.create_email_message(url, body, **header)
+
+    @abstractmethod
+    def _get_header_from_html(self, soup: BeautifulSoup) -> Dict[str, str]:
+        pass
+
+    @abstractmethod
+    def _get_body_from_html(
+        self, list_name: str, url: str, soup: BeautifulSoup
+    ) -> Union[str, None]:
+        pass
+
+    @staticmethod
+    def to_dict(msg: mboxMessage) -> Dict[str, List[str]]:
+        """Convert mboxMessage to a Dictionary"""
+        return MessageIO.to_dict(msg)
+
+    @staticmethod
+    def to_pandas_dataframe(msg: mboxMessage) -> pd.DataFrame:
+        """Convert mboxMessage to a pandas.DataFrame"""
+        return MessageIO.to_pandas_dataframe(msg)
+
+    @staticmethod
+    def to_mbox(msg: mboxMessage, filepath: str):
+        """
+        Parameters
+        ----------
+        msg : The Email.
+        filepath : Path to file in which the Email will be stored.
+        """
+        return MessageIO.to_mbox(msg, filepath)
+
+
+class AbstractList(ABC):
+    """
+    This class handles the scraping of a single mailing list of a public archive
+
+    Parameters
+    ----------
+    name : The of whom the list (e.g. 3GPP_COMMON_IMS_XFER, IEEESCO-DIFUSION, ...)
+    source : Contains the information of the location of the mailing list.
+        It can be either an URL where the list or a path to the file(s).
+    msgs : List of mboxMessage objects
+
+    Methods
+    -------
+    from_url()
+    from_messages()
+    from_mbox()
+    get_messages_from_url()
+    get_message_urls()
+    get_period_urls()
+    get_index_of_elements_in_selection()
+    to_dict()
+    to_pandas_dataframe()
+    to_mbox()
+    """
+
+    def __init__(
+        self,
+        name: str,
+        source: Union[List[str], str],
+        msgs: List[mboxMessage],
+    ):
+        self.name = name
+        self.source = source
+        self.messages = msgs
+
+    def __len__(self) -> int:
+        """Get number of messsages within the mailing list."""
+        return len(self.messages)
+
+    def __iter__(self):
+        """Iterate over each message within the mailing list."""
+        return iter(self.messages)
+
+    def __getitem__(self, index) -> mboxMessage:
+        """Get specific message at position `index` within the mailing list."""
+        return self.messages[index]
+
+    @classmethod
+    @abstractmethod
+    def from_url(
+        cls,
+        name: str,
+        url: str,
+        select: Optional[dict] = {"fields": "total"},
+        url_login: Optional[str] = None,
+        url_pref: Optional[str] = None,
+        login: Optional[Dict[str, str]] = {"username": None, "password": None},
+        session: Optional[requests.Session] = None,
+    ) -> "AbstractList":
```

Comment:
I'm confused about why the type hint for the output of this abstract method is a string "AbstractList" instead of a reference to the class.

Is this the best way to do it in Python?

Reply:
An answer can be found [here](https://stackoverflow.com/questions/40376641/type-hints-how-can-i-indicate-that-i-want-to-return-an-instance-of-the-actual-c). I have just adapted it after seeing other code.

Reply:
Ok, thanks. I was just ignorant about that.